hei.1.man
.nh
.TH hei 8

.SH NAME
.PP
hei - A simple consistent command wrapper for nix


.SH SYNOPSIS
.PP
hei

.PP
.RS

.nf
[--debug|-D]
[--dry-run|-d]
[--flake|-f]=[value]
[--help|-h]
[--version|-v]

.fi
.RE


.SH DESCRIPTION
.PP
A simple consistent command wrapper for nix

.PP
\fBUsage\fP:

.PP
.RS

.nf
hei [GLOBAL OPTIONS] [command [COMMAND OPTIONS]] [ARGUMENTS...]

.fi
.RE


.SH GLOBAL OPTIONS
.PP
\fB--debug, -D\fP: Enable debug logging

.PP
\fB--dry-run, -d\fP: Perform a dry run without making any changes

.PP
\fB--flake, -f\fP="": Path to flake. Will default to auto-detect

.PP
\fB--help, -h\fP: show help

.PP
\fB--version, -v\fP: print the version


.SH COMMANDS
.SH build
.PP
Run nix flake check on your flake

.PP
\fB--help, -h\fP: show help

.SS help, h
.PP
Shows a list of commands or help for one command

.SH check
.PP
Run checks on the given flake paths or the default ones if none are provided

.PP
\fB--help, -h\fP: show help

.SS help, h
.PP
Shows a list of commands or help for one command

.SH gc
.PP
Garbage collect & optimize nix store

.PP
\fB--all, -a\fP: Collect all garbage

.PP
\fB--help, -h\fP: show help

.PP
\fB--system, -s\fP: Collect system garbage

.SS help, h
.PP
Shows a list of commands or help for one command

.SH gen
.PP
Manage nix generations

.PP
\fB--help, -h\fP: show help

.SS list
.PP
List nix generations

.PP
\fB--help, -h\fP: show help

.SS help, h
.PP
Shows a list of commands or help for one command

.SS delete
.PP
Build the given flake paths or the default ones if none are provided

.PP
\fB--help, -h\fP: show help

.SS help, h
.PP
Shows a list of commands or help for one command

.SS diff
.PP
Build the given flake paths or the default ones if none are provided

.PP
\fB--help, -h\fP: show help

.SS help, h
.PP
Shows a list of commands or help for one command

.SS switch
.PP
Switch generation

.PP
\fB--help, -h\fP: show help

.SS help, h
.PP
Shows a list of commands or help for one command

.SS help, h
.PP
Shows a list of commands or help for one command

.SH p
.PP
Shortcut for nix profile commands

.PP
\fB--help, -h\fP: show help

.SS help, h
.PP
Shows a list of commands or help for one command

.SH rebuild
.PP
Rebuild your nix configuration

.PP
\fB--fast, -f\fP: Build in fast mode

.PP
\fB--help, -h\fP: show help

.PP
\fB--offline, -o\fP: Build in offline mode

.PP
\fB--rollback, -r\fP: Build in fast mode

.PP
\fB--update, -u\fP: Pull on nix flake before rebuilding

.SS help, h
.PP
Shows a list of commands or help for one command

.SH repl
.PP
open a repl in your nix config

.PP
\fB--help, -h\fP: show help

.SS help, h
.PP
Shows a list of commands or help for one command

.SH rollback
.PP
Roll back to previous generation of nixos. See gen list for the current generations.

.PP
\fB--help, -h\fP: show help

.SS help, h
.PP
Shows a list of commands or help for one command

.SH search
.PP
Search nixpkgs for packages

.PP
\fB--help, -h\fP: show help

.SS help, h
.PP
Shows a list of commands or help for one command

.SH show
.PP
Run nix flake show on the given flake paths or the default ones if none are provided

.PP
\fB--help, -h\fP: show help

.SS help, h
.PP
Shows a list of commands or help for one command

.SH ssh
.PP
Run a hei command on a remote NixOS system

.PP
\fB--help, -h\fP: show help

.SS help, h
.PP
Shows a list of commands or help for one command

.SH swap
.PP
Recursively swap nix-store symlinks with copies (or back)

.PP
\fB--help, -h\fP: show help

.SS help, h
.PP
Shows a list of commands or help for one command

.SH test
.PP
Run a nix check

.PP
\fB--help, -h\fP: show help

.PP
\fB--interactive, -i\fP: Run the test with the interctive test driver

.SS help, h
.PP
Shows a list of commands or help for one command

.SH upgrade
.PP
deprecated, use rebuild -u instead

.PP
\fB--help, -h\fP: show help

.SS help, h
.PP
Shows a list of commands or help for one command

.SH update
.PP
Update the given flake paths or the default ones if none are provided

.PP
\fB--help, -h\fP: show help

.SS help, h
.PP
Shows a list of commands or help for one command

.SH help, h
.PP
Shows a list of commands or help for one command

.SH completion
.PP
Generate shell completion scripts

.PP
\fB--help, -h\fP: show help

.SS help, h
.PP
Shows a list of commands or help for one command


CHANGELOG.md
##  (2025-11-25)

* chore: v0.2.1 ([7fe894f](https://code.bas.es/marcus/hei/commits/7fe894f))
* feat: add -c flag to show diff and confirm before switch ([05f17c5](https://code.bas.es/marcus/hei/commits/05f17c5))
* feat: add wrapper to ensure we have nom and nvd ([d5aaf7e](https://code.bas.es/marcus/hei/commits/d5aaf7e))
* fix: add missing command for diff ([1a144f7](https://code.bas.es/marcus/hei/commits/1a144f7))
* fix: use the right nom. :facepalm: ([67f7eea](https://code.bas.es/marcus/hei/commits/67f7eea))
* refactor: add stricter linter, make cli more testable, cleanup ([300b673](https://code.bas.es/marcus/hei/commits/300b673))
* refactor: rename exec command to make it more clear in captures stdio ([27e643e](https://code.bas.es/marcus/hei/commits/27e643e))



## 0.2.0 (2025-11-22)

* chore: prep for initial release ([10d3fcb](https://code.bas.es/marcus/hei/commits/10d3fcb))
* chore: setup gomod2nix ([2f0140c](https://code.bas.es/marcus/hei/commits/2f0140c))
* chore: start mocking up the commands ([6ce17d6](https://code.bas.es/marcus/hei/commits/6ce17d6))
* fix: align -u with rebuild, fix bad command ([449612b](https://code.bas.es/marcus/hei/commits/449612b))
* fix: restore original update functionality, add -p for pull ([c75b336](https://code.bas.es/marcus/hei/commits/c75b336))
* fix: restore switch as default rebuild command ([dd82933](https://code.bas.es/marcus/hei/commits/dd82933))
* feat: basic implementation of all the commands from nim version ([a8229d9](https://code.bas.es/marcus/hei/commits/a8229d9))
* feat: doc generation, update derivation with man and shell completion ([e8ae276](https://code.bas.es/marcus/hei/commits/e8ae276))
* feat: initial import ([1d361b0](https://code.bas.es/marcus/hei/commits/1d361b0))
* feat: start implementing some commands ([14c06ad](https://code.bas.es/marcus/hei/commits/14c06ad))





src/hei.nim
import std/[parseopt, os, envvars]

import strformat, strutils, sequtils
import hei/[commands, utils]

const version = staticExec("git describe --tags HEAD")

var flakePaths = @["/etc/nixos", "~/.config/nix-darwin", "~/.config/nix-config"]
if existsEnv("NIX_SYSTEM_FLAKE"):
  let envFlake = getEnv("NIX_SYSTEM_FLAKE")
  flakePaths.insert(@[envFlake])
var flakePath = findFlake(flakePaths)
if flakePath == "":
  echo "No system flake found in ", flakePaths
  echo "If your system flake is in a different location, set the NIX_SYSTEM_FLAKE environment variable"
  quit()

when isMainModule:
  var p = initOptParser()
  while true:
    p.next()
    case p.kind
    of cmdEnd:
      break
    of cmdShortOption, cmdLongOption:
      case p.key
      of "f", "flake":
        flakePath = p.val
      of "d", "dry-run":
        putEnv("NIX_DEBUG", "1")
      of "D", "debug":
        putEnv("NIX_DEBUG", "1")
        putEnv("NIX_SHOW_TRACE", "1")
      of "v", "version":
        echo fmt"hei {version} - running on {hostOs}({hostCpu})"
        quit()
      of "h", "help":
        break
      else:
        echo "Unknown option: ", p.key, ". run `hei` for help."
        quit()
    of cmdArgument:
      dispatchCommand(p.key, flakePath, p.remainingArgs)
      quit()

  echo "Error: No command specified."
  dispatchCommand("help", flakePath, @[])
